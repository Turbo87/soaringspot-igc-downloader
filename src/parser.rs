use html_escape::decode_html_entities;
use scraper::{Html, Selector};

#[derive(Debug, Clone)]
pub struct IgcFile {
    pub callsign: String,
    pub download_url: String,
}

pub fn parse_igc_files(html: &str) -> Result<Vec<IgcFile>, Box<dyn std::error::Error>> {
    let document = Html::parse_document(html);
    let mut igc_files = Vec::new();

    // Select all elements with data-content attributes containing download links
    let selector = Selector::parse(r#"[data-content*="download-contest-flight"]"#)?;

    for element in document.select(&selector) {
        // Extract the data-content attribute
        if let Some(data_content) = element.value().attr("data-content") {
            // Decode HTML entities and extract the download URL
            if let Some(download_url) = extract_download_url(data_content) {
                // Get the callsign from the text content of this cell
                let callsign = element.text().collect::<String>().trim().to_string();

                if !callsign.is_empty() {
                    igc_files.push(IgcFile {
                        callsign,
                        download_url,
                    });
                }
            }
        }
    }

    Ok(igc_files)
}

fn extract_download_url(data_content: &str) -> Option<String> {
    // The data_content contains HTML-encoded content
    // We need to decode it and extract the download URL with ?dl=1

    // Decode HTML entities using the html-escape library
    let decoded = decode_html_entities(data_content);

    // Parse the decoded HTML fragment using scraper
    let fragment = Html::parse_fragment(&decoded);

    // Select all links that contain "download-contest-flight" and have ?dl=1
    let link_selector = Selector::parse(r#"a[href*="download-contest-flight"][href*="dl=1"]"#)
        .expect("Invalid CSS selector");

    // Find the first matching link and extract its href
    fragment
        .select(&link_selector)
        .next()
        .and_then(|element| element.value().attr("href"))
        .map(|href| format!("https://www.soaringspot.com{}", href))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_igc_files_snapshot() {
        let html = include_str!("../tests/fixtures/day.html");
        let igc_files = parse_igc_files(html).expect("Failed to parse IGC files");

        insta::assert_debug_snapshot!(igc_files);
    }
}
