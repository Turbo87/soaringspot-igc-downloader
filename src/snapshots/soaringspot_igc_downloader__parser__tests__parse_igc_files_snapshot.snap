---
source: src/parser.rs
expression: igc_files
---
[
    IgcFile {
        callsign: "AM",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576293?dl=1",
    },
    IgcFile {
        callsign: "A3",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576289?dl=1",
    },
    IgcFile {
        callsign: "CP",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576272?dl=1",
    },
    IgcFile {
        callsign: "FL",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576287?dl=1",
    },
    IgcFile {
        callsign: "FLS",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576281?dl=1",
    },
    IgcFile {
        callsign: "AG",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576279?dl=1",
    },
    IgcFile {
        callsign: "194",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576280?dl=1",
    },
    IgcFile {
        callsign: "PP",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576312?dl=1",
    },
    IgcFile {
        callsign: "KT",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576306?dl=1",
    },
    IgcFile {
        callsign: "SN",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576309?dl=1",
    },
    IgcFile {
        callsign: "GE",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576311?dl=1",
    },
    IgcFile {
        callsign: "W",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576294?dl=1",
    },
    IgcFile {
        callsign: "B6",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576277?dl=1",
    },
    IgcFile {
        callsign: "CRN",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576295?dl=1",
    },
    IgcFile {
        callsign: "1A",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576296?dl=1",
    },
    IgcFile {
        callsign: "XI",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576298?dl=1",
    },
    IgcFile {
        callsign: "SF",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576290?dl=1",
    },
    IgcFile {
        callsign: "FY",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576292?dl=1",
    },
    IgcFile {
        callsign: "NW",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576310?dl=1",
    },
    IgcFile {
        callsign: "RSM",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576301?dl=1",
    },
    IgcFile {
        callsign: "AD",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576288?dl=1",
    },
    IgcFile {
        callsign: "IX",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576285?dl=1",
    },
    IgcFile {
        callsign: "MN",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576283?dl=1",
    },
    IgcFile {
        callsign: "B4",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576278?dl=1",
    },
    IgcFile {
        callsign: "KM",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576308?dl=1",
    },
    IgcFile {
        callsign: "Y3",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576271?dl=1",
    },
    IgcFile {
        callsign: "CX",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576303?dl=1",
    },
    IgcFile {
        callsign: "ZS",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576299?dl=1",
    },
    IgcFile {
        callsign: "YC",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576273?dl=1",
    },
    IgcFile {
        callsign: "LAB",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576274?dl=1",
    },
    IgcFile {
        callsign: "6L",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576284?dl=1",
    },
    IgcFile {
        callsign: "XW",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576305?dl=1",
    },
    IgcFile {
        callsign: "FP",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576286?dl=1",
    },
    IgcFile {
        callsign: "T5",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576302?dl=1",
    },
    IgcFile {
        callsign: "HR3",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576282?dl=1",
    },
    IgcFile {
        callsign: "Y2",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576300?dl=1",
    },
    IgcFile {
        callsign: "HG",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576276?dl=1",
    },
    IgcFile {
        callsign: "RZ",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576304?dl=1",
    },
    IgcFile {
        callsign: "44",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576297?dl=1",
    },
    IgcFile {
        callsign: "S5",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576275?dl=1",
    },
    IgcFile {
        callsign: "GM",
        download_url: "https://www.soaringspot.com/en_gb/download-contest-flight/5039-10179576291?dl=1",
    },
]
